pkg/controller/jobframework/reconciler.go

```diff
@@ -241,14 +241,16 @@ func (r *JobReconciler) ReconcileGenericJob(ctx context.Context, req ctrl.Reques
 
 	// 2. handle job is finished.
 	if condition, finished := job.Finished(); finished && wl != nil {
-		err := workload.UpdateStatus(ctx, r.client, wl, condition.Type, condition.Status, condition.Reason, condition.Message, constants.JobControllerName)
-		if err != nil {
-			log.Error(err, "Updating workload status")
-		}
 		// Execute job finalization logic
 		if err := r.finalizeJob(ctx, job); err != nil {
 			return ctrl.Result{}, err
 		}
+
+		err := workload.UpdateStatus(ctx, r.client, wl, condition.Type, condition.Status, condition.Reason, condition.Message, constants.JobControllerName)
```

Comment:
Consider the opposite case, where `finalizeJob` succeeds, and `UpdateStatus` fails. Would a subsequent reconcile get the workload updated correctly?

OTOH,
with the original code, why isn't `finalizeJob` called through `ensureOneWorkload`?

Reply:
>  with the original code, why isn't finalizeJob called through ensureOneWorkload?

We're calling `finalizeJob` only if there's no matching workload:

https://github.com/kubernetes-sigs/kueue/blob/main/pkg/controller/jobframework/reconciler.go#L421

Reply:
perhaps we should call finalize here? https://github.com/kubernetes-sigs/kueue/blob/f3bb6f2b6e509727e33b4aac84d2503cc9edaa0a/pkg/controller/jobframework/reconciler.go#L224

But let's check what happens in the new code if `UpdateStatus` fails.

Reply:
> But let's check what happens in the new code if UpdateStatus fails.

Added a test for that in 7535ac43.